### ISSUE TITLE
django-components no longer working

### ISSUE THREAD

**Issue by TheSteveBurgess** [carltongibson/django-template-partials#59]
Hi,

I’ve just upgraded to the latest version of django components and it’s no longer working with template partials.

Looking at the code django components is monkey patching Template and then calling the Template _render method:

https://github.com/EmilStenstrom/django-components/blob/594c0689bae2ae4729fca0c4d90abf3adab52733/src/django_components/component.py#L998

However, the TemplateProxy in template partials doesn’t implement an _render method. It directly calls self.nodelist.render(context), which is actually all the_render method does anyway.

https://github.com/carltongibson/django-template-partials/blob/694141c52c48e5e01609017b4ee6923e69df37a3/src/template_partials/templatetags/partials.py#L35

Would it be possible to bring the _render method in the TemplateProxy in line with the Django Template implementation? It would just need the _render method adding and the render method updating to point to it.

https://github.com/django/django/blob/5b4d949d7ca118e70985ffc53f8191b766591c12/django/template/base.py#L165

Happy to put a PR in if you’re OK with the changes 😊

**Comment by carltongibson** [comment]
Grrr. This is the trouble with using private API. We use render to avoid that

Did you track the change in components that caused this? Is there a tracking issue over there?

**Comment by carltongibson** [comment]
OK, pr was https://github.com/EmilStenstrom/django-components/pull/511. 

The commit for this was in https://github.com/EmilStenstrom/django-components/releases/tag/0.78 which was released in May (so like 6 months ago now). I'm curious this hasn't shown up before 🤔

TBH I'm not seeing (yet) why they don't inline the `_render()` call. It's not overridden in Django, so wondering why it's there. (Archeology needed.)

**Comment by carltongibson** [comment]
@TheSteveBurgess But, yes, I can't see that it would hurt. Happy to look at a PR[*]. Thanks. 


[*]: Trying to think about a meaningful test.

**Comment by TheSteveBurgess** [comment]
> The commit for this was in https://github.com/EmilStenstrom/django-components/releases/tag/0.78 which was released in May (so like 6 months ago now). I'm curious this hasn't shown up before 🤔

That thought did cross my mind. I made the changes locally and it fixed the issue. As far as I can tell the issue should 100% occur if you’re using a partial in a component (so setting template_name to something like “template.html#partialiwant”)

> [*]: Trying to think about a meaningful test.

We could mock _render and just confirm that it’s being called when render is called? Although that does sort of feel like creating a test for the sake of it…

Could it be a situation where its already covered by existing tests?

**Comment by JuroOravec** [comment]
Hey, off topic, but I just came across issue. If there's any more issue with django-components in the future, do mention me, so I can clarify or help.

**Comment by carltongibson** [comment]
Thanks @JuroOravec 🎁 

I'm planning to take this and do a release shortly. Any comments welcomed!

**Comment by JuroOravec** [comment]
So ofc I'm not familiar with internals of django-template-partials, but having a quick look at https://github.com/carltongibson/django-template-partials/pull/60/files#top:


- I think the critical part is here: https://github.com/carltongibson/django-template-partials/pull/60/files#diff-82e9847530faed28a95a84970c0ccc655c6e89f96ae5060036636c9a3382ceb8R32

- In django-components, we need to parametrize the call to `push_state()`, we do so [here](https://github.com/django-components/django-components/blob/0b65761fce857722b3f2c00e39a75f560214331f/src/django_components/util/django_monkeypatch.py#L97). It's necessary to be able to isolate the context of individual templates.

- So with respect to django-components, the ideal approach here would be to implement `TemplateProxy.render()` such that it internally calls `Template.render()`.

   E.g.
   ```py
   class TemplateProxy:
       ...
       def render(self, context):
           template = self.origin.loader.get_template(self.origin.template_name)
           return template.render(context)
   ```

**Comment by JuroOravec** [comment]
Btw, sorry that off topic again, but @carltongibson what's a good way to contact you? I came to `django-template-partials` because I thought I read somewhere that someone made a language server for `django-template-partials`, and I want to make one for django-components too. But I might be remembering wrong.

**Comment by carltongibson** [comment]
@JuroOravec Here's fine. 

Language server? That's gotta be @joshuadavidthomas

**Comment by JuroOravec** [comment]
Ah ok, yeah I messaged Josh too, because I know he made / is working on one, but I thought there was also someone else working on a different implementation. My bad!

**Comment by joshuadavidthomas** [comment]
> Ah ok, yeah I messaged Josh too, because I know he made / is working on one, but I thought there was also someone else working on a different implementation. My bad!

Yep, there is https://github.com/fourdigits/django-template-lsp as well.

**Comment by JuroOravec** [comment]
Ah yeah, that's the other one, thanks!

**Comment by JuroOravec** [comment]
Btw @carltongibson @joshuadavidthomas, cheeky question, but do you guys wanna help us building django-components (DJC)? DJC is already a superset of many other similar projects. My intention is to make DJC 1) into a standalone templatating engine that would supersede Django's vanilla engine, and 2) viable alternative to Vue and React in terms of features, ease of use, performance. DJC will allow e.g. literal lists and dicts in templates (e.g. `{{ [1, 2, 3 ] }}`) (already wrote a Rust parser for that), JS / CSS variables, scoped CSS, easy HTMX (HTML fragments) support, and a lot more.

**Comment by carltongibson** [comment]
@JuroOravec Certainly happy to watch from the stands. I have zero capacity to take on a new project though myself. 🤹‍♀️
